### PR TITLE
SFA-1: Integrate amplify and auth

### PR DESCRIPTION
**Pull request by nguyenhongphat0**
- [x] Integrate amplify
 - [x] Amplify add auth
 - [x] Login/Logout and navigation flow

### CODE REVIEW

File: src/app/app.module.ts [hunk 1]
```diff
@@ -23,9 +23,22 @@ import {
...
+  'aws_user_pools_web_client_id': '3c15dfd0cf8si85lpfj046lioo',
+  'oauth': {},
+  'aws_cloud_logic_custom': [
+      {
+          'name': 'AdminQueries',
+          'endpoint': 'https://xaeo1xy4fc.execute-api.ap-southeast-1.amazonaws.com/dev',
+          'region': 'ap-southeast-1',
+      },
+  ],
+});
```
Comment (peterbean410) [inline src/app/app.module.ts:41]: Let's not hard code config parameters here, instead put them in variable environments.

File: e2e/src/app.po.ts [hunk 2]
```diff
@@ -8,4 +8,49 @@ export class AppPage {
...
+        return submatches;
+      };
+
+      let matches = findAllMatches(selectors.shift().trim(), [opt_parentElement || document], true);
+      while (selectors.length > 0 && matches.length > 0) {
+        matches = findAllMatches(selectors.shift().trim(), matches, false);
+      }
+      return matches;
+    });
+  }
```
Comment (peterbean410) [inline e2e/src/app.po.ts:55]: I would put these in `auth.po.ts` file

File: e2e/src/app.e2e-spec.ts [hunk 3]
```diff
@@ -1,17 +1,39 @@
...
+  it('should login success with correct credential', async () => {
+    await page.navigateTo();
+    const username: ElementFinder = await page.getUsernameInput();
+    await username.sendKeys('phatnh');
+    const password: ElementFinder = await page.getPasswordInput();
+    await password.sendKeys('12345678');
+    await password.sendKeys(Key.ENTER);
+    await browser.wait(ExpectedConditions.presenceOf(element(by.css('ngx-pages'))), 10000);
+    expect(await page.getTitleText()).toEqual('Users Management');
   });
```
Comment (peterbean410) [inline e2e/src/app.e2e-spec.ts]: We are testing AUTH flow here, so let's put them under `auth.e2e-spec.ts`

File: .gitignore [hunk 4]
```diff
@@ -39,3 +39,23 @@ testem.log
...
 Thumbs.db
+
+#amplify
+amplify/\#current-cloud-backend
+amplify/.config/local-*
+amplify/logs
+amplify/mock-data
+amplify/backend/amplify-meta.json
+amplify/backend/awscloudformation
+amplify/backend/.temp
```
Comment (peterbean410) [inline .gitignore:50]: how about `amplify/` ? We want to ignore all amplify related config, don't we?

File: package.json [hunk 5]
```diff
@@ -23,6 +23,7 @@
     "lint:ci": "npm run lint && npm run lint:styles",
     "pree2e": "webdriver-manager update --standalone false --gecko false",
     "e2e": "ng e2e",
+    "protractor": "protractor",
```
Comment (peterbean410) [inline package.json:26]: Can you explain why we need this?

File: e2e/src/app.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+  it('should login success with correct credential', async () => {
+    await page.navigateTo();
+    const username: ElementFinder = await page.getUsernameInput();
+    await username.sendKeys('phatnh');
+    const password: ElementFinder = await page.getPasswordInput();
+    await password.sendKeys('12345678');
+    await password.sendKeys(Key.ENTER);
+    await browser.wait(ExpectedConditions.presenceOf(element(by.css('ngx-pages'))), 10000);
+    expect(await page.getTitleText()).toEqual('Users Management');
```
Comment (peterbean410) [inline e2e/src/app.e2e-spec.ts:36]:
```suggestion
    expect(await page.getTitleText()).toEqual('User Management');
```

File: src/app/pages/pages-routing.module.ts [hunk 7]
```diff
@@ -2,75 +2,20 @@ import { RouterModule, Routes } from '@angular/router';
...
-      loadChildren: () => import('./tables/tables.module')
-        .then(m => m.TablesModule),
-    },
-    {
-      path: 'miscellaneous',
-      loadChildren: () => import('./miscellaneous/miscellaneous.module')
-        .then(m => m.MiscellaneousModule),
+      path: 'users',
+      loadChildren: () => import('./users/users.module')
+        .then(m => m.UsersModule),
```
Comment (peterbean410) [inline src/app/pages/pages-routing.module.ts:14]: Should change all UsersModule to the singular form UserModule

File: src/app/pages/pages.module.ts [hunk 8]
```diff
@@ -3,18 +3,16 @@ import { NbMenuModule } from '@nebular/theme';
...
+import { UsersModule } from './users/users.module';
 
 @NgModule({
   imports: [
     PagesRoutingModule,
     ThemeModule,
     NbMenuModule,
-    DashboardModule,
-    ECommerceModule,
+    UsersModule,
```
Comment (peterbean410) [inline src/app/pages/pages.module.ts:15]: Please ask me about the noun to use before you start putting the name to the code. Should be UserModule

File: src/app/pages/users/users.component.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+@Component({
+  selector: 'ngx-users',
+  template: `<router-outlet></router-outlet>`,
+})
+export class UsersComponent implements OnInit {
+
+  constructor(private titleService: Title) { }
+
+  ngOnInit() {
+    this.titleService.setTitle('Users Management');
```
Comment (peterbean410) [inline src/app/pages/users/users.component.ts:13]:
```suggestion
    this.titleService.setTitle('User Management');
```

File: src/app/pages/users/users.module.ts [hunk 10]
```diff
@@ -13,12 +13,12 @@ import { FsIconComponent } from './tree-grid/tree-grid.component';
     NbIconModule,
     NbInputModule,
     ThemeModule,
-    TablesRoutingModule,
+    UsersRoutingModule,
```
Comment (peterbean410) [inline src/app/pages/users/users.module.ts:16]:
```suggestion
    UserRoutingModule,
```

File: src/app/pages/users/users.module.ts [hunk 11]
```diff
@@ -13,12 +13,12 @@ import { FsIconComponent } from './tree-grid/tree-grid.component';
...
     Ng2SmartTableModule,
   ],
   declarations: [
     ...routedComponents,
     FsIconComponent,
   ],
 })
-export class TablesModule { }
+export class UsersModule { }
```
Comment (peterbean410) [inline src/app/pages/users/users.module.ts:24]:
```suggestion
export class UserModule { }
```

File: src/app/pages/users/users-routing.module.ts [hunk 12]
```diff
@@ -17,17 +17,22 @@ const routes: Routes = [{
...
 @NgModule({
   imports: [RouterModule.forChild(routes)],
   exports: [RouterModule],
 })
-export class TablesRoutingModule { }
+export class UsersRoutingModule { }
 
 export const routedComponents = [
-  TablesComponent,
+  UsersComponent,
```
Comment (peterbean410) [inline src/app/pages/users/users-routing.module.ts:35]:
```suggestion
  UserListComponent,
```

**Review by peterbean410 — CHANGES_REQUESTED** [review]
Plural -> singular at some places
1. UsersModule -> UserModule
2. UsersComponent -> UserListComponent
3. Users Management -> User Management
4. UsersRoutingModule -> UserRoutingModule
5. users.component.ts -> user-list.component.ts

And also look at my other comments, pleaseeee.

File: package.json (hunk 5 above)
Comment (nguyenhongphat0) [reply package.json:26]: `npm run e2e` first compile the source code and start a dev server before execute e2e test, and it's very slow. If we already have a dev server running (by using `npm start`), `npm run protractor` will execute the e2e test right a way. It's save lots of times when developing e2e test, boss 😫

**Comment by nguyenhongphat0** [comment]
Woohoo 🥳

**Pull request merged**